torch/distributed/fsdp/wrap.py

```diff
@@ -195,21 +178,32 @@ def _module_wrap_policy(
     return isinstance(module, tuple(module_classes))
 
 
-class ModuleWrapPolicy(_FSDPPolicy):
-    """This is a wrapper around :func:`_module_wrap_policy`."""
+class ModuleWrapPolicy(_WrapPolicy):
+    """
+    This policy wraps every module of the specified module classes, passing in
+    the kwargs given to the root.
+    """
 
     def __init__(self, module_classes: Iterable[Type[nn.Module]]):
         module_classes_set = set(module_classes)
-        self._policy: Callable = functools.partial(
-            _module_wrap_policy,
-            module_classes=module_classes_set,
-        )
         self._module_classes = module_classes_set
         self._module_classes_str = str(module_classes_set)
 
-    @property
-    def policy(self):
-        return self._policy
+    def _run_policy(
+        self,
+        root_module: nn.Module,
+        ignored_modules: Set[nn.Module],
+        root_kwargs: Dict[str, Any],
+    ) -> Dict[nn.Module, Dict[str, Any]]:
+        module_classes = tuple(self._module_classes)
+        target_module_to_kwargs: Dict[nn.Module, Dict[str, Any]] = {}
+        for module in root_module.modules():
+            if module in ignored_modules:
+                continue
+            elif isinstance(module, module_classes):
+                # Shallow copy to avoid coupling changes across modules
+                target_module_to_kwargs[module] = copy.copy(root_kwargs)
```

Comment:
why do we need to copy the args? Do we expect root_kwargs to change?

Reply:
We must copy the kwargs or else if a function overrides it (e.g. the mixed precision overrides), then it will override it for _every_ target module. We have to avoid them sharing a reference to the same kwarg dict.